src/components/tl/ucp/tl_ucp_ep.c

```diff
@@ -29,47 +30,54 @@ static inline ucc_status_t ucc_tl_ucp_connect_ep(ucc_tl_ucp_context_t *ctx,
 ucc_status_t ucc_tl_ucp_connect_ctx_ep(ucc_tl_ucp_context_t *ctx, ucc_rank_t ctx_rank)
 {
     return ucc_tl_ucp_connect_ep(ctx, &ctx->eps[ctx_rank],
-                                 ctx->addr_storage->addresses,
+                                 (char*)ctx->addr_storage->addresses,
                                  ctx->addr_storage->max_addrlen, ctx_rank);
 }
 
-ucc_status_t ucc_tl_ucp_connect_team_ep(ucc_tl_ucp_team_t *team, ucc_rank_t team_rank)
+ucc_status_t ucc_tl_ucp_connect_team_ep(ucc_tl_ucp_team_t *team, ucc_rank_t team_rank,
+                                        ucc_context_id_t key, ucp_ep_h *ep)
 {
     ucc_tl_ucp_context_t *ctx = UCC_TL_UCP_TEAM_CTX(team);
-    if (team->context_ep_storage) {
+    ucc_status_t status;
+    if (ctx->eps) {
         int ctx_rank = -1; //TODO map to ctx rank
         ucc_assert(0);
-        return ucc_tl_ucp_connect_ctx_ep(ctx, ctx_rank);
+        status = ucc_tl_ucp_connect_ctx_ep(ctx, ctx_rank);
+        *ep = ctx->eps[0]; //TODO map
+        return status;
     }
-    return ucc_tl_ucp_connect_ep(ctx, &team->eps[team_rank],
-                                 team->addr_storage->addresses,
-                                 team->addr_storage->max_addrlen, team_rank);
+    status = ucc_tl_ucp_connect_ep(ctx, ep,
+                                   (char*)team->addr_storage->addresses,
+                                   team->addr_storage->max_addrlen, team_rank);
+    if (UCC_OK == status) {
+        tl_ucp_hash_put(ctx->ep_hash, key, *ep);
+    }
+    return status;
 }
 
-ucc_status_t ucc_tl_ucp_close_eps(ucc_tl_ucp_context_t *ctx, ucp_ep_h *eps,
-                                  ucc_rank_t n_eps)
+ucc_status_t ucc_tl_ucp_close_eps(ucc_tl_ucp_context_t *ctx)
 {
     ucc_tl_ucp_ep_close_state_t *state = &ctx->ep_close_state;
-    ucs_status_t status;
+    ucp_ep_h                     ep;
+    ucs_status_t                 status;
     if (state->close_req) {
         ucp_worker_progress(ctx->ucp_worker);
         status = ucp_request_check_status(state->close_req);
         if (status != UCS_OK) {
             return UCC_INPROGRESS;
         }
         ucp_request_free(state->close_req);
-        eps[state->ep++] = NULL;
     }
-
-    for (; state->ep < n_eps; state->ep++) {
-        if (!eps[state->ep]) {
-            continue;
+    while (1) {
+        ep = tl_ucp_hash_pop(ctx->ep_hash);
+        if (NULL == ep) {
+            break;
         }
         state->close_req =
-            ucp_ep_close_nb(eps[state->ep], UCP_EP_CLOSE_MODE_FLUSH);
+            ucp_ep_close_nb(ep, UCP_EP_CLOSE_MODE_FLUSH);
         if (UCS_PTR_IS_ERR(state->close_req)) {
             tl_error(ctx->super.super.lib, "failed to start ep close, ep %p",
-                     eps[state->ep]);
```

Comment:
looks like ep field is not needed in ucc_tl_ucp_ep_close_state_t  anymore, should we delete it?